src/components/Sessions/RequireSession.js

```diff
@@ -1,42 +1,77 @@
-import React from 'react';
+import React, {useEffect, useReducer} from 'react';
 import PropTypes from 'prop-types';
 import {Link} from 'react-router-dom';
 import {FormattedMessage} from 'react-intl';
 
 import Anchor from 'components/Anchor';
 import Card from 'components/Card';
 import ErrorMessage from 'components/ErrorMessage';
+import Modal from "components/modals/Modal";
 
-// import SessionExpiry from './SessionExpiry';
+const WARN_SESSION_TIMEOUT_FACTOR = 0.9; // once 90% of the session expiry time has passed, show a warning
 
+const untilWarningMilliseconds = (expiryDate) => {
+  /*
+  Calculate the time until the session expiry warning time should be displayed.
+   */
+  return (expiryDate - new Date()) * WARN_SESSION_TIMEOUT_FACTOR;
+};
+
+const RequireSession = ({expired = false, expiryDate = null, children}) => {
+  const [expiryWarning, setExpiryWarning] = useReducer(
+    (draft, action) => ({...draft, ...action}),
+    {display: false, accepted: false});
+
+  useEffect(() => {
+    if (expiryDate) {
+      const ms = untilWarningMilliseconds(expiryDate);
+      const timeout = setTimeout(() => {
+        setExpiryWarning({display: true});
+      }, ms);
+      return () => clearTimeout(timeout);
+    }
```

Comment:
there might be an edge case here where the component is unmounted but there are still stale references and `setExpiryWarning` may still be called. Though that should probably clear up with the `clearTimeout`.

I also suspect that multiple hook calls can be scheduled to set the warning display to true for every render of this component where it does not get unmounted :thinking: